### DISCUSSION TITLE
Question about IPXE crosscert generation

### DISCUSSION THREAD

**Discussion by andy191x** [ipxe/ipxe#1362] [General]
Hello!

Is the crosscert generation portion of iPXE open source? Referring to the certs located at ( https://ca.ipxe.org/auto/ ).

I've poked around a bit through the ipxe repos, but didn't see anything that explicitly generates this folder.

Thanks!

**Comment by NiKiZe** [comment]
Not at this time, not because of it being any kind of secret, but more about the scripts themselves not being in a sharable state. See #126 and maybe #127

**Reply by andy191x** [reply]
Thanks for the reply! It seems that #126 could be relevant, but that one has been on the backburner for a while.

A little backstory from my end. I've inherited an environment that uses a custom iPXE build for booting with an in-house root CA. A previous engineer was able to determine how to generate the crosscert ".der" files against the new root. However, his implementation doesn't generate an exact 1:1 to the [official cross signed certs](https://ca.ipxe.org/auto/). I'm not sure if this has skewed over time, or if we never had a fully compatible implementation. The end result is random trust failures when connecting to some domains via HTTPS.

Do you have any options for sharing the official crosscert generation scripts? If you'd prefer, I'd be happy to move discussion to a private channel.

**Reply by NiKiZe** [reply]
There have been other discussions on this as well, but maybe in your case the best way forward could be the Customer section at https://ipxe.org/contact

And to clarify, I'm just a user and helping out where I can, and not deep in to the iPXE development.

**Reply by andy191x** [reply]
Will do, thank you for the help!

**Comment by joernheissler** [comment]
Hi,
`validator.c` is where the magic happens. There's a comment `Enter certificateSet`, and that's what those cross signed files are: `CMS CertificateSet`. See e.g. RFC 5652.

Each file is a set of three certs in exactly this order:
* root cert of some public CA, cross signed by ipxe's intermediate CA.
* ipxe's intermediate CA cert, signed by ipxe's root CA.
* ipxe's root CA, self-signed.

The latter two certs are the same across all cross-certifications.
The actual root cert is heavily modified. It only retains subject and subject-public-key from the original, which is sufficient for cross signing.

The ipxe root cert's sha256 sum is referenced in `rootcert.c`, completing the chain of trust.

`CertificateSet` can be decoded to the more common X.509/PEM files with this script:
```python
#!/usr/bin/python3
# License: CC0

from pathlib import Path
from sys import argv
from asn1crypto.cms import CertificateSet
from asn1crypto.pem import armor

for idx, cert in enumerate(CertificateSet.load(Path(argv[1]).read_bytes())):
    Path(f"{argv[1]}-{idx}.crt").write_bytes(armor("CERTIFICATE", cert.chosen.dump()))
```

Encoding is basically the reverse of that
```python
#!/usr/bin/python3
# License: CC0

from pathlib import Path
from sys import argv
from zlib import crc32
from asn1crypto.core import SequenceOf
from asn1crypto.pem import unarmor
from asn1crypto.x509 import Certificate

class OrderedCertificateSet(SequenceOf):
    tag = 17
    _child_spec = Certificate

certs = [Certificate.load(unarmor(Path(filename).read_bytes())[2]) for filename in argv[1:]]
crc = 0xffffffff - crc32(certs[0]["tbs_certificate"]["subject"].dump())
Path(f"{crc:08x}.der").write_bytes(OrderedCertificateSet(certs).dump())
```

I decoded + re-encoded all 149 cross signed certs and the result is exactly the input.

Now all you need to do is create your own certs. That can be done with standard tools like openssl.

n.b.: Above code violates X.690 Section 11.6, but so does ipxe's own cross cert generation script. I haven't checked the ipxe code if this might cause your randomish validation issues. Another reason for your problems might be https servers that fail to serve the intermediate cert.

**Reply by andy191x** [reply]
This is very helpful, thank you for sharing!

**Reply by mcb30** [reply]
> `CertificateSet` can be decoded to the more common X.509/PEM files with this script:
> 
> ```python
> #!/usr/bin/python3
> # License: CC0
> 
> from pathlib import Path
> from sys import argv
> from asn1crypto.cms import CertificateSet
> from asn1crypto.pem import armor
> 
> for idx, cert in enumerate(CertificateSet.load(Path(argv[1]).read_bytes())):
>     Path(f"{argv[1]}-{idx}.crt").write_bytes(armor("CERTIFICATE", cert.chosen.dump()))
> ```
> 
> Encoding is basically the reverse of that
> 
> ```python
> #!/usr/bin/python3
> # License: CC0
> 
> from pathlib import Path
> from sys import argv
> from zlib import crc32
> from asn1crypto.core import SequenceOf
> from asn1crypto.pem import unarmor
> from asn1crypto.x509 import Certificate
> 
> class OrderedCertificateSet(SequenceOf):
>     tag = 17
>     _child_spec = Certificate
> 
> certs = [Certificate.load(unarmor(Path(filename).read_bytes())[2]) for filename in argv[1:]]
> crc = 0xffffffff - crc32(certs[0]["tbs_certificate"]["subject"].dump())
> Path(f"{crc:08x}.der").write_bytes(OrderedCertificateSet(certs).dump())
> ```
> 
> I decoded + re-encoded all 149 cross signed certs and the result is exactly the input.

This is great: thank you for this.  I've been meaning to rewrite the crosscert generation script into Python for years, but never found the time to check which underlying Python libraries would be best to use to get small and neat code.

> n.b.: Above code violates X.690 Section 11.6, but so does ipxe's own cross cert generation script. I haven't checked the ipxe code if this might cause your randomish validation issues. Another reason for your problems might be https servers that fail to serve the intermediate cert.

I never knew that rule for DER encoding, thank you for mentioning it!  It explains why several standards seem to use a SEQUENCE when the underlying information is logically a SET: it avoids the need to sort the values.

iPXE won't care about the ordering: it will just parse all the certificates in the set and then use the issuer names to build up the chain (starting with the issuer name it's trying to find in the first place, which is the cross-signed CA's subject name).  So you should find that your cross-signed files still work if you reorder the SET elements correctly.  I'm guessing that doing so may also allow to you simplify the Python code further, since you wouldn't need to hack a SequenceOf to use a SetOf tag?

**Reply by joernheissler** [reply]
> I'm guessing that doing so may also allow to you simplify the Python code further, since you wouldn't need to hack a SequenceOf to use a SetOf tag?

Yeah, you can just use `from asn1crypto.cms import CertificateSet` and `CertificateSet(certs)` instead. Saves another few lines of code ^_^

**Comment by sinuscosinustan** [comment]
We've wrote an Ansible role + Python script that cross-signs all installed Root CAs of a system: https://github.com/hetznercloud/ansible-role-ipxe-ca
